### DISCUSSION TITLE
Incompatible with OpenComposite (OpenOVR)?

### DISCUSSION THREAD

**Discussion by SilkReyn** [Rectus/openxr-steamvr-passthrough#13] [Q&A] [answered]
Did a one-shot test of a SteamVR game that is forwarded to OpenXR by using OpenComposite (individual installation in app directory). Was expecting to access this OpenXR layer that way, but the game crashed on start. So I guess this passthrough is incompatible with OpenComposite, because both call openvr_api.dll?

**Comment by Rectus** [comment]
That's likely the issue. OpenComposite replaces openvr_api.dll with it's own version that translates calls into OpenXR, while this API layer uses openvr_api.dll to access the SteamVR camera and overlay features. 

I haven't tested it myself, but if you are using the system-wide install of OpenComposite, you could try turning it off and installing it manually into the game directory. Make sure you only replace the .dll in the game directory and not the one in the directory the API layer itself is installed in. The API Layer is supposed to specifically load and use the version of openvr_api.dll in the same directory as the layer .dll itself, so it might still work even if the game is running OpenComposite.

**Comment by Rectus** [comment, marked as answer]
I looked a bit at it myself, and while manually setting up OpenComposite didn't work, recompiling the layer with OpenVR statically linked worked. It only renders the passthrough in one eye though, which I think is a bug in the layer. I'll see if I can fix it.

**Reply by SilkReyn** [reply]
Ooh thats great news and a good sign that it could possibly work together. Strange though, that dynamic linking didn't work as expected even though you supply openvr within the installation. Maybe the original or OC's lib gets loaded with the game first and then remains in the execution cache. Would static linking then a problem as a workaround? If you had it distributed with a particular version already and use only a basic part of it, was there a reason behind? By the way, do you have an assembly of the binary with statically linked API?
Now I hope that the mono-eye doesn't turn out to be too deep of a rabbit hole...

**Reply by Rectus** [reply]
Here's the version with static linking.

[openxr-steamvr-passthrough_static_linking.zip](https://github.com/user-attachments/files/16769360/openxr-steamvr-passthrough_static_linking.zip)

**Comment by Rectus** [comment]
I released an update that should let it work properly with OpenComposite, please report any games it still doesn't work with.
https://github.com/Rectus/openxr-steamvr-passthrough/releases/tag/v0.2.8-beta

**Reply by SilkReyn** [reply]
Thanks alot for preparing me that build to test with OC. With it I was then able to get into the game and see your overlay menu, but passthrough (override) wasn't rendered, not even one-eyed. In debug log everything looked as it would happily do its thing, aside from temporal shaders failing, but I had that feature disabled. Also in the first static_linking dll, the swapchain might have select wrong index, using index1 array size 1. Compared with Beatsaber, which as native OpenXR game worked perfectly with your plugin, had index0 array size 2.
By the way, is it normal, that the device in debug menu is only recognized when pressing refresh button during runtime?

Ya anyways, passthrough via OpenComposite does not work in my games (Unity Engine, might be around 2018.3.11), also not with beta0.2.8. Either no camera rendered ingame, game does hang in starting up or crash when starting up. One game shows an OpenComposite error dialog: dx11compositor.cpp:309 XR_ERROR_CALL_ORDER_INVALID xrWaitSwapchainImage
I can however play the games via OC when Passthrough is disabled in OpenXR layers (and steamVR is restarted).

Just one more question, when I say OpenComposite, do we both mean this https://gitlab.com/znixian/OpenOVR ?
Because I just discovered other sources on Github which may be forks from other people...

I'm sad that it doesn't work for my games, but I understand that this is an out-of-spec use case not really designed for. I'm impressed of your work and what you're archived. And also the time you're on it by now, recently I've seen your posts on Unreal forums from years ago...

**Reply by Rectus** [reply]
Thanks!

It's normal for the debug menu devices to be empty, I didn't want to scan all the properties needlessly.

The XR_ERROR_CALL_ORDER_INVALID is interesting, I think the cause might be that I delay releasing swapchain images until I've rendered onto them, which means there might be a race condition if the application starts rendering multiple frames before finishing the first one. My implementation does go against the OpenXR spec, but I haven't noticed it affecting any native games. I'll have to check it out with some Unity games.

And yeah, that's the version of OpenComposite I used for testing.

**Reply by Rectus** [reply]
I tested out some Unity games, but could not get any of them running with OpenComposite, regardless of if the API layer was installed. The games start, but fail to submit any frames. I was testing with a Valve Index and SteamVR Beta 2.8.4.

Have you checked if OpenComposite works on your end without the layer installed?

**Reply by SilkReyn** [reply]
Sure, checked 2 games running with OC before installing yours. Wouldn't bothered going on without that working. After installation of the XR layer it still can be tested individually by disabling the layer in OpenVR options and restarting VR. I also have Valve Index, but on stable branch.

Back then I had problems downloading OC's last build and looked for alternative sources. Took a version from 1st June [DL mirror](https://opencomposite.znix.xyz/builds/view_build?commit=43e551a4506880ab1a71b8b9fec2fd7fbb27372f) found in forum [dont update opencomposite](https://forum.il2sturmovik.com/topic/88670-dont-update-opencomposite/) where they report problems after a specific update. Manually replaced openvr_api.dll in Game/game_data/Plugins without using the Switcher App. Replaced openvr_api.dll was created year 2020. I assume more actual Unity games might fail because they going to have their own OpenXR module, like Beat Saber (UnityOpenXR.dll).

**Reply by Rectus** [reply]
Thanks, the older version works, and I can replicate the issue. 

Looking at it through RenderDoc, it is backface culling my triangles for some reason. Flipping it to cull the front faces makes it work, but breaks all other applications. I'll have to see if I can figure out why it's culling the triangles.

**Reply by Rectus** [reply]
Figured it out. Unity renders with a flipped vertical axis, and OpenComposite seems to flip the FoV vertically in a way the native Unity OpenXR plugin doesn't.

**Reply by SilkReyn** [reply]
Good idea to look at the native OpenXR plugin for reference! I was thinking that Unity eventually use a different orientation of the fundamental forward Z-Axis. But when I understand you correctly, the problem lies now in OC how to treat Unity geometry? I admit that I haven't checked if they have Unity games in their supported list. Though how come that OC alone runs my games well enough...apparently?

**Reply by Rectus** [reply]
Basically OpenXR provides a set of FoV values to render to, but applications can render at any FoV (including flipping the values) and pass the values actually used to render back to the runtime. OC seems to flip the value to match ow Unity rendered it, so it makes the runtime flip the rendered view back. This is all allowed by the OpenXR spec. 

Since I have to render the passthrough in the same perspective as  the game rendered the frame at, I have to use the values the game and OC returns, and it's impossible to cover all eventualities.

**Reply by SilkReyn** [reply]
Thank you for the explanation. Could an user configurable option help, that decides if one specific game need the flip? Or is it more like a mixed case where within same game some views are flipped others not?

**Reply by Rectus** [reply]
I can detect it from checking the sign on the FoV values I get back. I've already added a fix to the code.

**Reply by SilkReyn** [reply]
I forgot to ask about the other problem I got with the layer, did you see a XR_ERROR_CALL_ORDER_INVALID on game startup, something with the swapchain failed? I think I get this when I start the (Unity) game where I previously had setup forced overlay.

**Reply by Rectus** [reply]
That's possibly an issue with my swapchain handling, although I can't say for sure. I haven't been able to replicate it myself.
